### PR TITLE
fix Issue 18397 - Poor implementation of std.conv.hexString results i…

### PR DESCRIPTION
**Pull request by WalterBright**
…n unintended bloat

The trick is to make it an enum, use a mixin string literal rather than an array, and instantiate the template in std.conv rather than in user code.

For:
```
import std.conv;
void test() {
    auto data = cast(ubyte[]) hexString!"deadbeef";
}
```
now compiles to:
```
___a4_deadbeef	comdat
	db	0deh,0adh,0beh,0efh,000h	;.....
___a4_deadbeef	ends
_D5test24testFZv	comdat
	assume	CS:_D5test24testFZv
		mov	ECX,offset FLAT:___a4_deadbeef
		mov	EAX,4
		ret
_D5test24testFZv	ends
```
instead of the hundreds of bytes of code it used to.

### CODE REVIEW

**Comment by WalterBright** [comment]
```
std/digest/md.d(483): Error: cannot implicitly convert expression `"\xc3\xfc\xd3\xd7a\x92\xe4\x00}\xfbIl\xcag\xe1;"` of type `string` to `ubyte[16]`
```
I knew somebody was likely misusing hexString :-(

File: std/digest/crc.d [hunk 1]
```diff
@@ -381,7 +381,7 @@ struct CRC(uint N, ulong P) if (N == 32 || N == 64)
                     "1234567890123456789012345678901234567890");
     assert(digest == cast(ubyte[]) hexString!"724aa97c");
 
-    enum ubyte[4] input = hexString!"c3fcd3d7";
+    enum ubyte[4] input = cast(ubyte[4]) hexString!"c3fcd3d7";
```
Comment (wilzbach) [inline std/digest/crc.d]: If we decide the breakage is worth it, it should at least be mentioned in the changelog.

File: std/digest/crc.d [hunk 2]
```diff
@@ -421,7 +421,7 @@ struct CRC(uint N, ulong P) if (N == 32 || N == 64)
                          "1234567890123456789012345678901234567890");
     assert(digest == cast(ubyte[]) hexString!"bd3eb7765d0a22ae");
 
-    enum ubyte[8] input = hexString!"c3fcd3d7efbeadde";
+    enum ubyte[8] input = cast(ubyte[8]) hexString!"c3fcd3d7efbeadde";
```
Comment (carun) [inline std/digest/crc.d]: Just wondering if it is mandatory to use `enum ubyte[8] input`? `auto input` would look idiomatic.

File: std/digest/crc.d (hunk 1 above)
Comment (carun) [inline std/digest/crc.d]: Just wondering if it is mandatory to use `enum ubyte[8] input`? `auto input` would look idiomatic.

File: std/digest/crc.d [hunk 3]
```diff
@@ -461,7 +461,7 @@ struct CRC(uint N, ulong P) if (N == 32 || N == 64)
                         "1234567890123456789012345678901234567890");
     assert(digest == cast(ubyte[]) hexString!"760cd2d3588bf809");
 
-    enum ubyte[8] input = hexString!"c3fcd3d7efbeadde";
+    enum ubyte[8] input = cast(ubyte[8]) hexString!"c3fcd3d7efbeadde";
```
Comment (carun) [inline std/digest/crc.d]: Just wondering if it is mandatory to use `enum ubyte[8] input`? `auto input` would look idiomatic.

File: std/digest/ripemd.d [hunk 4]
```diff
@@ -649,7 +649,7 @@ struct RIPEMD160
                     "1234567890123456789012345678901234567890");
     assert(digest == cast(ubyte[]) hexString!"9b752e45573d4b39f4dbd3323cab82bf63326bfb");
 
-    enum ubyte[20] input = hexString!"f71c27109c692c1b56bbdceb5b9d2865b3708dbc";
+    enum ubyte[20] input = cast(ubyte[20]) hexString!"f71c27109c692c1b56bbdceb5b9d2865b3708dbc";
```
Comment (carun) [inline std/digest/ripemd.d]: Just wondering if it is mandatory to use `enum ubyte[16] input`? `auto input` would look idiomatic.

File: std/digest/sha.d [hunk 5]
```diff
@@ -1117,7 +1117,7 @@ alias SHA512_256 = SHA!(1024, 256); /// SHA alias for SHA-512/256, hash is ubyte
     assert(digest512_224 == cast(ubyte[]) hexString!"37ab331d76f0d36de422bd0edeb22a28accd487b7a8453ae965dd287");
     assert(digest512_256 == cast(ubyte[]) hexString!"9a59a052930187a97038cae692f30708aa6491923ef5194394dc68d56c74fb21");
 
-    enum ubyte[20] input = hexString!"a9993e364706816aba3e25717850c26c9cd0d89d";
+    enum ubyte[20] input = cast(ubyte[20]) hexString!"a9993e364706816aba3e25717850c26c9cd0d89d";
```
Comment (carun) [inline std/digest/sha.d]: Just wondering if it is mandatory to use `enum ubyte[20] input`? `auto input` would look idiomatic.

**Comment by JackStouffer** [comment]
@carun When you have a suggestion that applies to multiple pieces of code, simply stating something like "same throughout" will suffice.

And it must be at least `enum input`, as `auto` turns it into a variable.

**Comment by WalterBright** [comment]
> What are those non-hex chars doing in there?

The string-to-ascii conversion represents non-printable characters as \xnn

**Comment by carun** [comment]
@JackStouffer I see, thanks!

**Comment by schveiguy** [comment]
> The string-to-ascii conversion represents non-printable characters as \xnn

OK, so it produces the correct string literal, but then when displaying the error message shows the printable ones as the actual chars.

I was looking at the code, and couldn't figure out of something weird was sneaking in!

**Comment by WalterBright** [comment]
added changelog entry

**Comment by WalterBright** [comment]
oops, pushed wrong button

**Comment by WalterBright** [comment]
For reference, the documentation: https://dlang.org/phobos/std_conv.html#hexString

File: changelog/fix18397.dd [hunk 6]
```diff
@@ -1,18 +1,18 @@
...
-generated object file
-* the generated object file no longer contains references to TypeInfo
-and various druntime functions
-* it is now compatible with -betterC mode
+$(UL
+  $(LI behavior consistency with the documentation)
+  $(LI the `hexString` template instantiations no longer appear in the generated object file)
+  $(LI the generated object file no longer contains references to `TypeInfo` and various druntime functions)
+  $(LI it is now compatible with `-betterC` mode)
+)
```
Comment (wilzbach) [inline changelog/fix18397.dd]: FYI @WalterBright even if `changed.d` overlays some goodies over Ddoc (no need for the ugly `$(P)`, Ddoc is still used in the background :/

**Comment by wilzbach** [comment]
@WalterBright FYI: I "DDocified" your changelog entry. The changelog generator still uses old Ddoc behind the scenes.

**Pull request merged**